fee-hook/src/FeeHook.sol

```diff
@@ -0,0 +1,208 @@
+                bytes memory payload = abi.encode();
+                // FIXME tune this value, could push it lower
+                uint128 messageFee = 3_000_000_000_000_000;
+                uint128 GAS_LIMIT = 500000;
+                uint128 VALUE = uint128(address(this).balance);
+                bytes memory options = OptionsBuilder.newOptions().addExecutorLzReceiveOption(GAS_LIMIT, VALUE - messageFee);
+                // Note that external payable functions cannot be called directly
+                // within a contract, so we have to instantiate a new instance
+                // of the FeeHook contract to call the sendCrossChain function.
+                FeeHook(payable(address(this))).sendCrossChain{value: VALUE}(
+                    payload,
+                    options,
+                    VALUE,
+                    address(this)
+                );
+            }
+            uncollectedFees[poolId] = 0;
+
+        }
+        return (BaseHook.afterSwap.selector, 0);
+    }
+
+    function _beforeAddLiquidity(
+        address,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal override returns (bytes4) {
+        PoolId poolId = key.toId();
+        if (TOKEN_ID[poolId] == 0) {
+            TOKEN_ID[poolId] = posm.nextTokenId() - 1;
+            poolIds.push(PoolId.unwrap(poolId));
+        }
+        return BaseHook.beforeAddLiquidity.selector;
+    }
+
+    // Allows the owner to transfer an LP position to another address
+    // in case any manual intervention is needed.
+    function transferPosition(
+        address to,
+        bytes32 poolId) external onlyOwner
+    {
+        PoolId id = PoolId.wrap(poolId);
+        uint256 tokenId = TOKEN_ID[id];
+        require(tokenId != 0, "Token ID not found");
+        transferredPoolIds.push(poolId);
+        PositionManager(payable(address(posm))).safeTransferFrom(address(this), to, tokenId, "");
+    }
+
+
+    function withdrawFees() external {
+        if (block.chainid != destinationChainId) {
+            revert("Not on destination chain. Cannot withdraw fees");
+        }
+        uint256 totalSupply = IERC20(vMooneyAddress).totalSupply();
+        require(totalSupply > 0, "No vMooney supply");
+        uint256 userBalance = IERC20(vMooneyAddress).balanceOf(msg.sender);
+        uint256 userProportion = userBalance * 1e18 / totalSupply; // Multiply by 1e18 to preserve precision
+        uint256 allocated = (userProportion * totalReceived) / 1e18; // Divide by 1e18 to normalize
+        require(allocated > 0, "Nothing to withdraw");
+        uint256 withdrawnByUser = totalWithdrawnPerUser[msg.sender];
+        uint256 withdrawable = allocated - withdrawnByUser;
+        require(withdrawable > 0, "Nothing to withdraw");
+
+        totalWithdrawnPerUser[msg.sender] += withdrawable;
+        totalWithdrawn += withdrawable;
+        transferETH(msg.sender, withdrawable);
```

Comment:
What if we record balances increases every swap, but allow people to withdraw (i.e. transfer ETH) in a separate function? This gets them to "check in" with MoonDAO every now and then + have that "feel good" moment where they see their rewards accruing over time

Reply:
added the view.